### PR TITLE
Use cURL for faster restores to development

### PR DESCRIPTION
**Pull request by geoffharcourt**
We had reports of slow restores from backups larger than 500MB in #94.

This change uses `curl` rather than `heroku pg:pull` to perform a
database restore. We download the remote backup to `tmp`, re-initialize
the development database, perform a restore with `pg_restore` and then
delete the tempfile.

h/t to @croaky for the original idea and @calebthompson for figuring out which command-line arguments to use.

### CODE REVIEW

**Comment by croaky** [comment]
Nice! Have you done any tests to see speed differences? Want me to try some large datasets and report back?

**Comment by geoffharcourt** [comment]
@croaky you can't install the executables via Rubygems any longer, so the easiest way to test is to install the development build via Homebrew:

https://github.com/thoughtbot/parity/releases/tag/development. Be sure to verify that you don't have an rbenv shim intercepting your attempts to run `production`, `staging`, or `development.

File: lib/parity/backup.rb [hunk 1]
```diff
@@ -34,21 +34,32 @@ def restore_from_development
...
+    def wipe_development_database
+      Kernel.system("dropdb #{development_db} && createdb #{development_db}")
     end
 
-    def pull_remote_database_to_development
+    def download_remote_backup
       Kernel.system(
-        "heroku pg:pull DATABASE_URL #{development_db} --remote #{from} "\
-          "#{additional_args}",
+        "curl -o tmp/latest.backup $(#{from} pg:backups public-url -q)",
```
Comment (bernerdschaefer) [inline lib/parity/backup.rb]: I think the `$(#{from} ...)` should be quoted, as there's no reason the URL might not contain, e.g., `&` or other nasty shell characters.

Comment (geoffharcourt) [inline lib/parity/backup.rb]: @bernerdschaefer like this?

```ruby
"curl -o tmp/latest.backup '$(#{from} pg:backups public-url -q)'",
```

Comment (bernerdschaefer) [inline lib/parity/backup.rb]: They might have to be double quotes to execute the `$()`. Heroku's docs don't quote the URL (https://devcenter.heroku.com/articles/heroku-postgres-backups#downloading-your-backups), however, so maybe it's fine as is.

Comment (geoffharcourt) [inline lib/parity/backup.rb]: @bernerdschaefer quoted in d4f0348. Thank you very much for pointing out this risk.

**Comment by bernerdschaefer** [comment]
👍

**Comment by geoffharcourt** [comment]
@croaky I'm going to wait for your report against the DBs that caused you trouble before merging.

**Comment by fagiani** [comment]
@geoffharcourt I can confirm that even for pretty small databases, using `curl` approach is way faster than `heroku pg:pull`. That is probably true because retrieving the dump file and running `pg_restore` from a local source is much faster than retrieving the data from the remote source as it gets restored. Hope that helps!

**Comment by geoffharcourt** [comment]
@fagiani thank you for the info! I am going to try to merge this soon.

**Comment by geoffharcourt** [comment]
I'm encountering some issues in testing where the database role in Heroku are named in such a way that they cause permission errors when the restore step happens. I'm looking in to what we can do about that.

**Comment by geoffharcourt** [comment]
`time development restore staging` with v0.10.0 for a 650 MB database:

```
 557.00 real         5.22 user         6.27 sys
```

`time development restore staging` with #100 for a 650 MB database:

```
 47.16 real         2.21 user         1.59 sys
```

**Comment by geoffharcourt** [comment]
@croaky can you test on a database today with the development release? https://github.com/thoughtbot/parity/releases/tag/development

**Comment by fagiani** [comment]
hey @geoffharcourt, how long it usually takes to get a release on brew formulae after the PR is merged? Thanks again!

**Comment by geoffharcourt** [comment]
Hi @fagiani at the moment it's driven by my available time! If you want to install the development build on Homebrew (using the `--devel` switch it's at this commit). I'd appreciate the testing feedback to confirm that everything is working as expected.

**Comment by fagiani** [comment]
@geoffharcourt maybe I am missing something but tried those commands without success:
```
$ brew tap thoughtbot/formulae
$ brew install --devel parity
Warning: thoughtbot/formulae/parity-0.10.0 already installed
$ brew install --devel parity-0.10.1
Error: No available formula with the name "thoughtbot/formulae/parity-0.10.1"
```

Any clues?

**Comment by geoffharcourt** [comment]
@fagiani try uninstalling parity, then execute `brew install --devel parity` with no argument? You may have to delete your Homebrew cache.

**Comment by fagiani** [comment]
@geoffharcourt Thanks for the direction. It worked!

Tested both restoring from `staging` and from `production` on a couple of heroku apps. Results are incomparable!

I'd guess using cURL will get all data compressed at once then process it as batch locally while the former approach will end up doing several requests to the remote endpoint for the various restore steps and as you can see most of it is network activity that can get even worse with latency.

Here is a sample of a pretty small database (~35MB):

```
0.10.0: development restore-from production  1.15s user 0.70s system 0% cpu 4:34.71 total
0.10.1: development restore-from production  0.85s user 0.31s system 8% cpu 13.521 total
```
Keep Rocking!

**Pull request merged**